src/index.ts

```diff
@@ -147,80 +164,106 @@ module.exports = (app: ChartProviderApp): Plugin => {
-      : resolveUniqueChartPaths(props.chartPaths, configBasePath)
+      : resolveUniqueChartPaths(config.chartPaths, configBasePath)
 
-    const onlineProviders = _.reduce(
-      props.onlineChartProviders,
+    onlineProviders = _.reduce(
+      config.onlineChartProviders,
       (result: { [key: string]: object }, data) => {
         const provider = convertOnlineProviderConfig(data)
         result[provider.identifier] = provider
         return result
       },
       {}
     )
+
+    chartPaths.forEach((p) => {
+      app.debug('watching folder..', p)
+      watchers.push(fs.watch(p, 'utf8', () => handleWatchEvent()))
+    })
+
     app.debug(
       `Start charts plugin. Chart paths: ${chartPaths.join(
         ', '
       )}, online charts: ${Object.keys(onlineProviders).length}`
     )
 
-    // Do not register routes if plugin has been started once already
-    pluginStarted === false && registerRoutes()
-    pluginStarted = true
-    const urlBase = `${app.config.ssl ? 'https' : 'http'}://localhost:${
-      'getExternalPort' in app.config ? app.config.getExternalPort() : 3000
-    }`
-    app.debug('**urlBase**', urlBase)
-    app.setPluginStatus('Started')
+    return loadCharts()
+  }
+
+  // Load chart files
+  const loadCharts = async () => {
+    app.debug(`Loading Charts....`)
 
-    const loadProviders = bluebird
-      .mapSeries(chartPaths, (chartPath: string) => findCharts(chartPath))
-      .then((list: ChartProvider[]) =>
-        _.reduce(list, (result, charts) => _.merge({}, result, charts), {})
+    try {
+      const plist = await bluebird.mapSeries(chartPaths, (chartPath: string) =>
+        findCharts(chartPath)
       )
+      const charts = _.reduce(
+        plist,
+        (result, charts) => _.merge({}, result, charts),
+        {}
+      )
+      app.debug(
+        `Chart plugin: Found ${
+          _.keys(charts).length
+        } charts from ${chartPaths.join(', ')}.`
+      )
+      chartProviders = _.merge({}, charts, onlineProviders)
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    } catch (e: any) {
+      console.error(`Error loading chart providers`, e.message)
+      chartProviders = {}
+      app.setPluginError(`Error loading chart providers`)
+    }
+  }
 
-    return loadProviders
-      .then((charts: { [key: string]: ChartProvider }) => {
-        app.debug(
-          `Chart plugin: Found ${
-            _.keys(charts).length
-          } charts from ${chartPaths.join(', ')}.`
-        )
-        chartProviders = _.merge({}, charts, onlineProviders)
-      })
-      .catch((e: Error) => {
-        console.error(`Error loading chart providers`, e.message)
-        chartProviders = {}
-        app.setPluginError(`Error loading chart providers`)
-      })
+  const refreshProviders = async () => {
+    const td = Date.now() - (lastWatchEvent as number)
+    app.debug(`last watch event time elapsed = ${td}`)
+    if (lastWatchEvent && td > 5000) {
```

Comment:
Did I understand the logic correctly: if we notice a change in some chart data we reload the chart data the first time the data is needed after 5 seconds of the change?

Could we not just mark the data as stale and when it is next needed reload it? Or always reload when a change is noticed? Why the 5 second delay?

Reply:
This is due to the way `FSWatcher` produces events and the ability to distinguish when a:
1. New file is created
2. An exisiting file is replaced
3. An existing file is renamed
4. Multiple files are being added / deleted

There are multiple events produced during a simple file rename with no discernable metadata to indicate the actual operation taking place.... so the delay is to ensure the stream of events has ceased to indicate the process has completed. 
Without it we would trigger multiple reloads while a rename, etc. is in process and throw exceptions.